batch.js

```diff
@@ -0,0 +1,317 @@
+                    type: GraphQLString,
+                  },
+                  [inputFieldName]: pgField(
+                    build,
+                    fieldWithHooks,
+                    inputFieldName,
+                    {
+                      description: `The \`${inflection.pluralize(
+                        tableTypeName,
+                      )}\` that was upserted by this mutation.`,
+                      type: new GraphQLList(Table),
+                    },
+                    {
+                      // isPgCreatePayloadResultField: true,
+                      pgFieldIntrospection: table,
+                    },
+                  ),
+                }
+              },
+            },
+            {
+              __origin: `Adding table upsert batch payload type for ${describePgEntity(
+                table,
+              )}. You can  disable the built-in create mutation via:\n\n  ${sqlCommentByAddingTags(
+                table,
+                { omit: 'create' },
+              )}`,
+              isMutationPayload: true,
+              // isPgCreatePayloadType: true,
+              pgIntrospection: table,
+            },
+          )
+
+          // Create upsert fields from each introspected table
+          const fieldName = `upsert${tableTypeName}Batch`
+          memo = build.extend(
+            memo,
+            {
+              [fieldName]: fieldWithHooks(
+                fieldName,
+                context => {
+                  const { getDataFromParsedResolveInfoFragment } = context
+                  return {
+                    description: `Upserts a batch of \`${inflection.pluralize(
+                      tableTypeName,
+                    )}\`.`,
+                    type: PayloadType,
+                    args: {
+                      input: {
+                        type: new GraphQLNonNull(InputType),
+                      },
+                    },
+                    async resolve(data, { input }, { pgClient }, resolveInfo) {
+                      const parsedResolveInfoFragment = parseResolveInfo(
+                        resolveInfo,
+                      )
+                      const resolveData = getDataFromParsedResolveInfoFragment(
+                        parsedResolveInfoFragment,
+                        PayloadType,
+                      )
+                      const upsertedRowsAlias = sql.identifier(Symbol())
+                      const query = queryFromResolveData(
+                        upsertedRowsAlias,
+                        upsertedRowsAlias,
+                        resolveData,
+                        {},
+                      )
+                      const inputData = input[inputFieldName]
+
+                      // no inputs, no upserts
+                      if (!inputData || !inputData.length) {
+                        return {
+                          clientMutationId: input.clientMutationId,
+                          data: [],
```

Comment:
I think you need `query: build.$$isQuery` Here to make the query field on the mutation payload work?

Reply:
Do you mean to the `!inputData.length` return, or to the root of the `extends` options like in [MutationPayloadQueryPlugin](https://github.com/graphile/graphile-engine/blob/master/packages/graphile-build/src/plugins/MutationPayloadQueryPlugin.js#L25)? I'm adding the inner non-null constraint, so I guess the `!inputData.length` return will be superfluous

Reply:
I meant both clientMutationId/data return objects

Reply:
just tested, seems to work without it:
```graphql
upsertEventBatch(input: $EventBatchInput) {
    events {
      entityId
    }
    query {
    	allEvents {
        nodes { entityId }
      }
    }
  }
```

Reply:
👍 Perfect. I couldn't remember if it was needed or not.